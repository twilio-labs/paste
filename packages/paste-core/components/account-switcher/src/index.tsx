export * from "./AccountSwitcher";
export * from "./AccountSwitcherGroup";
export * from "./AccountSwitcherSeparator";
export * from "./AccountSwitcherItem";
export * from "./AccountSwitcherItemRadio";
export * from "./AccountSwitcherBadge";
export * from "./useAccountSwitcher";
