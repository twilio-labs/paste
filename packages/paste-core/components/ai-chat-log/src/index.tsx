export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageActionGroup } from "./AIChatMessageActionGroup";
export type { AIChatMessageActionGroupProps } from "./AIChatMessageActionGroup";
export { AIChatMessageBody } from "./AIChatMessageBody";
export type { AIChatMessageBodyProps } from "./AIChatMessageBody";
export { AIChatMessageAuthor } from "./AIChatMessageAuthor";
export type { AIChatMessageAuthorProps } from "./AIChatMessageAuthor";
export { AIChatMessageActionCard } from "./AIChatMessageActionCard";
export type { AIChatMessageActionCardProps } from "./AIChatMessageActionCard";
export { AIChatMessageLoading } from "./AIChatMessageLoading";
export type { AIChatMessageLoadingProps } from "./AIChatMessageLoading";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger, AIPartialIDChat, AIChat } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";

export type { AIMessageVariants } from "./AIMessageContext";
