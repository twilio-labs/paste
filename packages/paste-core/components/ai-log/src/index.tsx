export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageMeta } from "./AIChatMessageMeta";
export type { AIChatMessageMetaProps } from "./AIChatMessageMeta";
export { AIChatMessageMetaItem } from "./AIChatMessageMetaItem";
export type { AIChatMessageMetaItemProps } from "./AIChatMessageMetaItem";
export { AIChatMessageContent } from "./AIChatMessageContent";
export type { AIChatMessageContentProps } from "./AIChatMessageContent";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";
