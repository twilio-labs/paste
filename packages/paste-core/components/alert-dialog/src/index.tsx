import * as React from 'react';
import * as PropTypes from 'prop-types';
import {useUID} from '@twilio-paste/uid-library';
import {useTransition} from '@twilio-paste/animation-library';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';
import {ModalDialogOverlay} from '@twilio-paste/modal';
import {AlertDialogHeader} from './AlertDialogHeader';
import {AlertDialogBody} from './AlertDialogBody';
import {AlertDialogContent} from './AlertDialogContent';
import {AlertDialogFooter} from './AlertDialogFooter';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getAnimationStates = (): any => ({
  from: {opacity: 0, transform: `scale(0.675)`},
  enter: {opacity: 1, transform: `scale(1)`},
  leave: {opacity: 0, transform: `scale(0.675)`},
  // https://www.react-spring.io/docs/hooks/api
  config: {
    mass: 0.5,
    tension: 370,
    friction: 26,
  },
});

export interface AlertDialogProps extends React.HTMLAttributes<HTMLElement>, Pick<BoxProps, 'element'> {
  children: NonNullable<React.ReactNode>;
  destructive?: boolean;
  heading: string;
  isOpen: boolean;
  onConfirm: () => void;
  onConfirmLabel: string;
  onDismiss: () => void;
  onDismissLabel: string;
}

export const AlertDialog = React.forwardRef<HTMLDivElement, AlertDialogProps>(
  (
    {
      children,
      destructive,
      element = 'ALERT_DIALOG',
      heading,
      isOpen,
      onConfirm,
      onConfirmLabel,
      onDismiss,
      onDismissLabel,
      ...props
    },
    ref
  ) => {
    const transitions = useTransition(isOpen, getAnimationStates());
    const headingID = useUID();
    const bodyID = useUID();

    return (
      <>
        {transitions(
          (styles, item) =>
            item && (
              <ModalDialogOverlay isOpen={isOpen} style={{opacity: styles.opacity}}>
                <Box
                  // @ts-expect-error Render overlay as box for customization
                  as={AlertDialogContent}
                  {...safelySpreadBoxProps(props)}
                  aria-labelledby={headingID}
                  aria-describedby={bodyID}
                  element={element}
                  ref={ref}
                  role="alertdialog"
                  style={styles}
                >
                  <AlertDialogHeader headingID={headingID} element={`${element}_HEADER`}>
                    {heading}
                  </AlertDialogHeader>
                  <AlertDialogBody bodyID={bodyID} element={`${element}_BODY`}>
                    {children}
                  </AlertDialogBody>
                  <AlertDialogFooter
                    destructive={destructive}
                    element={`${element}_FOOTER`}
                    onDismiss={onDismiss}
                    onDismissLabel={onDismissLabel}
                    onConfirm={onConfirm}
                    onConfirmLabel={onConfirmLabel}
                  />
                </Box>
              </ModalDialogOverlay>
            )
        )}
      </>
    );
  }
);

AlertDialog.displayName = 'AlertDialog';

AlertDialog.propTypes = {
  children: PropTypes.node.isRequired,
  destructive: PropTypes.bool,
  element: PropTypes.string,
  heading: PropTypes.string.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onConfirm: PropTypes.func.isRequired,
  onConfirmLabel: PropTypes.string.isRequired,
  onDismiss: PropTypes.func.isRequired,
  onDismissLabel: PropTypes.string.isRequired,
};
