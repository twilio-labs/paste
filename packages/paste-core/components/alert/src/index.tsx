import * as React from 'react';
import * as PropTypes from 'prop-types';
import {ValueOf} from '@twilio-paste/types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import {MediaObject, MediaFigure, MediaBody} from '@twilio-paste/media-object';
import {Button} from '@twilio-paste/button';
import {CloseIcon} from '@twilio-paste/icons/esm/CloseIcon';
import {ErrorIcon} from '@twilio-paste/icons/esm/ErrorIcon';
import {InformationIcon} from '@twilio-paste/icons/esm/InformationIcon';
import {WarningIcon} from '@twilio-paste/icons/esm/WarningIcon';

export const AlertRoles = {
  ERROR: 'alert',
  NEUTRAL: 'status',
  WARNING: 'alert',
} as const;
export const AlertVariants = {
  ERROR: 'error',
  NEUTRAL: 'neutral',
  WARNING: 'warning',
} as const;
export const AlertBackgroundColors = {
  ERROR: 'colorBackgroundErrorLightest',
  NEUTRAL: 'colorBackgroundNeutralLightest',
  WARNING: 'colorBackgroundWarningLightest',
} as const;
export const AlertBorderColors = {
  ERROR: 'colorBorderErrorLight',
  NEUTRAL: 'colorBorderNeutralLight',
  WARNING: 'colorBorderWarningLight',
} as const;

export type AlertVariants = ValueOf<typeof AlertVariants>;
export type AlertBackgroundColors = ValueOf<typeof AlertBackgroundColors>;
export type AlertBorderColors = ValueOf<typeof AlertBorderColors>;
export type AlertRoles = ValueOf<typeof AlertRoles>;

export interface AlertProps {
  id?: never;
  className?: never;
  children: NonNullable<React.ReactNode>;
  onDismiss?: () => void;
  role?: string;
  variant: AlertVariants;
}

const renderAlertIcon = (variant: AlertVariants): React.ReactElement => {
  switch (variant) {
    case AlertVariants.ERROR:
      return <ErrorIcon iconColor="colorTextErrorDark" decorative={false} title="error: " size="sizeIcon30" />;
    case AlertVariants.WARNING:
      return <WarningIcon iconColor="colorTextWarningDark" decorative={false} title="warning: " size="sizeIcon30" />;
    case AlertVariants.NEUTRAL:
    default:
      return <InformationIcon iconColor="colorTextIcon" decorative={false} title="information: " size="sizeIcon30" />;
  }
};

const Alert: React.FC<AlertProps> = ({children, onDismiss, variant, role, ...props}) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      backgroundColor={AlertBackgroundColors[variant.toUpperCase()]}
      borderColor={AlertBorderColors[variant.toUpperCase()]}
      borderBottomWidth="borderWidth20"
      borderBottomStyle="solid"
      paddingLeft="space70"
      paddingRight="space70"
      paddingTop="space40"
      paddingBottom="space30"
      role={role != null ? role : AlertRoles[variant.toUpperCase()]}
    >
      <MediaObject as="div">
        <MediaFigure as="div" spacing="space40">
          {renderAlertIcon(variant)}
        </MediaFigure>
        <MediaBody as="div">{children}</MediaBody>
        {onDismiss && typeof onDismiss === 'function' && (
          <MediaFigure align="end" spacing="space70">
            <Button onClick={onDismiss} variant="link" size="reset">
              <CloseIcon iconColor="colorTextIcon" decorative={false} title="dismiss this alert" size="sizeIcon30" />
            </Button>
          </MediaFigure>
        )}
      </MediaObject>
    </Box>
  );
};
Alert.displayName = 'Alert';

if (process.env.NODE_ENV === 'development') {
  Alert.propTypes = {
    children: PropTypes.node.isRequired,
    onDismiss: PropTypes.func,
    role: PropTypes.string,
    variant: PropTypes.oneOf(Object.keys(AlertVariants).map(variant => AlertVariants[variant])).isRequired,
  };
}
export {Alert};
