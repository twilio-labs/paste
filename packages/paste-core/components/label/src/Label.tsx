import * as React from 'react';
import * as PropTypes from 'prop-types';
import {styled, css} from '@twilio-paste/styling-library';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import {Flex} from '@twilio-paste/flex';
import {Text} from '@twilio-paste/text';
import {ScreenReaderOnly} from '@twilio-paste/screen-reader-only';
import {TextColor} from '@twilio-paste/style-props';

export interface LabelProps extends React.LabelHTMLAttributes<HTMLLabelElement> {
  as?: 'label' | 'legend';
  children: NonNullable<React.ReactNode>;
  disabled?: boolean;
  htmlFor: string | undefined;
  marginBottom?: 'space0';
  required?: boolean;
  variant?: 'default' | 'inverse';
}

// eslint-disable-next-line emotion/syntax-preference
const StyledRequiredDot = styled(Text)(
  css({
    backgroundColor: 'colorBackgroundRequired',
    borderRadius: '50%',
    cursor: 'pointer',
    display: 'block',
    height: '4px',
    marginRight: 'space20',
    width: '4px',
  })
);

const Label: React.FC<LabelProps> = ({as, marginBottom, required, disabled, children, variant, ...props}) => {
  let textColor = 'colorText' as TextColor;
  if (disabled && variant === 'inverse') {
    textColor = 'colorTextInverseWeak';
  } else if (disabled) {
    textColor = 'colorTextWeak';
  } else if (variant === 'inverse') {
    textColor = 'colorTextInverse';
  }
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      as={as}
      // Setting a bottom border here to prevent Bootstrap's bottom border
      // on legend in Console.
      borderBottomWidth="borderWidth0"
      display="block"
      marginBottom={marginBottom || 'space20'}
      paddingLeft="space0"
      paddingRight="space0"
      textTransform="none"
    >
      <Flex as="span" vAlignContent="center">
        {required ? (
          <StyledRequiredDot as="span" lineHeight="lineHeight30">
            <ScreenReaderOnly>Required: </ScreenReaderOnly>
          </StyledRequiredDot>
        ) : null}
        <Text
          as="span"
          fontSize="fontSize30"
          fontWeight="fontWeightSemibold"
          lineHeight="lineHeight30"
          color={textColor}
          cursor={disabled ? 'not-allowed' : 'pointer'}
        >
          {children}
        </Text>
      </Flex>
    </Box>
  );
};

Label.displayName = 'Label';

Label.defaultProps = {
  as: 'label',
};

if (process.env.NODE_ENV === 'development') {
  Label.propTypes = {
    as: PropTypes.oneOf(['label', 'legend']),
    disabled: PropTypes.bool,
    htmlFor: PropTypes.string,
    marginBottom: PropTypes.oneOf(['space0']),
    required: PropTypes.bool,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    variant: PropTypes.oneOf(['default', 'inverse']) as any,
  };
}

export {Label};
