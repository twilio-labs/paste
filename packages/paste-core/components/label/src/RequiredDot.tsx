import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';
import {ScreenReaderOnly} from '@twilio-paste/screen-reader-only';

export interface RequiredDotProps {
  element?: BoxProps['element'];
}

export const RequiredDot: React.FC<RequiredDotProps> = ({element = 'REQUIRED_DOT', ...props}) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      alignItems="center"
      justifyContent="center"
      as="span"
      cursor="pointer"
      display="flex"
      height="sizeIcon30"
      element={`${element}_WRAPPER`}
    >
      <Box
        as="span"
        backgroundColor="colorBackgroundRequired"
        borderRadius="borderRadiusCircle"
        display="block"
        height="4px"
        width="4px"
        element={element}
      >
        <ScreenReaderOnly>Required: </ScreenReaderOnly>
      </Box>
    </Box>
  );
};

RequiredDot.displayName = 'RequiredDot';

RequiredDot.propTypes = {
  element: PropTypes.string,
};
