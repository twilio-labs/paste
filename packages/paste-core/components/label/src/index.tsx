export * from './Label';
export * from './RequiredDot';
