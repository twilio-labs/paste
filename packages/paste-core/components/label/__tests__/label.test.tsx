import * as React from 'react';
import {render, screen} from '@testing-library/react';
import {Theme} from '@twilio-paste/theme';
import {Label} from '../src';

describe('Label render', () => {
  it('it should render', (): void => {
    const {asFragment} = render(
      <Theme.Provider theme="console">
        <Label htmlFor="input">child</Label>
      </Theme.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('it should render with required', (): void => {
    const {asFragment} = render(
      <Theme.Provider theme="console">
        <Label htmlFor="input" required>
          child
        </Label>
      </Theme.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Label for prop', () => {
  const initialProps = {
    htmlFor: 'input',
  };

  it('should have a for prop on label', () => {
    render(
      <Label data-testid="test-label" {...initialProps}>
        Label
      </Label>
    );
    expect(screen.getByTestId('test-label').getAttribute('for')).toEqual('input');
  });
});

describe('Label required prop', () => {
  const initialProps = {
    htmlFor: 'input',
    required: true,
  };

  it('should have a required indicator', () => {
    render(<Label {...initialProps}>label</Label>);
    expect(screen.getByText('Required:')).not.toBeNull();
  });
});
