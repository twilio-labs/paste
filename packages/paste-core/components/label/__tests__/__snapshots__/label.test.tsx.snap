// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label render it should render 1`] = `
<DocumentFragment>
  .emotion-3 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-2 {
  box-sizing: border-box;
  border-bottom-width: 0;
  display: block;
  margin-bottom: 0.25rem;
  padding-left: 0;
  padding-right: 0;
  text-transform: none;
}

.emotion-1 {
  box-sizing: border-box;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-width: 0;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  font-size: 0.875rem;
  font-weight: 500;
  line-height: 1.25rem;
  color: rgb(40,42,43);
  cursor: pointer;
}

<div
    class="emotion-3"
  >
    <label
      class="emotion-2"
      display="block"
      for="input"
    >
      <span
        class="emotion-1"
        display="flex"
      >
        <span
          class="emotion-0"
          color="colorText"
          cursor="pointer"
          font-size="fontSize30"
          font-weight="fontWeightSemibold"
        >
          child
        </span>
      </span>
    </label>
  </div>
</DocumentFragment>
`;

exports[`Label render it should render with required 1`] = `
<DocumentFragment>
  .emotion-5 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-4 {
  box-sizing: border-box;
  border-bottom-width: 0;
  display: block;
  margin-bottom: 0.25rem;
  padding-left: 0;
  padding-right: 0;
  text-transform: none;
}

.emotion-3 {
  box-sizing: border-box;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-width: 0;
}

.emotion-2 {
  margin: 0;
  padding: 0;
  font-size: 0.875rem;
  font-weight: 500;
  line-height: 1.25rem;
  color: rgb(40,42,43);
  cursor: pointer;
}

.emotion-1 {
  box-sizing: border-box;
  background-color: rgb(222,88,88);
  border-radius: 50%;
  cursor: pointer;
  display: block;
  height: 4px;
  line-height: 1.25rem;
  margin-right: 0.25rem;
  width: 4px;
}

.emotion-0 {
  position: absolute;
  margin: -1px;
  border: 0;
  padding: 0;
  width: 1px;
  height: 1px;
  overflow: hidden;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  text-transform: none;
  white-space: nowrap;
}

<div
    class="emotion-5"
  >
    <label
      class="emotion-4"
      display="block"
      for="input"
    >
      <span
        class="emotion-3"
        display="flex"
      >
        <span
          class="emotion-1"
          cursor="pointer"
          display="block"
          height="4px"
          width="4px"
        >
          <span
            class="emotion-0"
          >
            Required: 
          </span>
        </span>
        <span
          class="emotion-2"
          color="colorText"
          cursor="pointer"
          font-size="fontSize30"
          font-weight="fontWeightSemibold"
        >
          child
        </span>
      </span>
    </label>
  </div>
</DocumentFragment>
`;
